fix(publish): fix version conflict recognition for github packages

petermetz opened the pull request:
## Description, Motivation and Context

The version conflict recognition logic was not working for github
because it returns a different `code` for the same error compared to
what npmjs.org returns.

I've added two utility functions to check for the error code and
message to make the code more readable in the error handler of the
publish operation.

This way there is a robust detection supporting both npmjs.com and
npm.pkg.github.com registries, not just npmjs.org.

Signed-off-by: Peter Somogyvari <peter.metz@unarin.com>

## How Has This Been Tested?

I've added a new test case within the file
`packages/publish/src/__tests__/publish-from-git.spec.ts`
to cover specifically the GitHub npm registry's error code,
so now the test verifies both npmjs.org and github.com registries
error codes not just npmjs.org as before.

I've also tested it manually, on a mono-repo I maintain where
there is a mixture of github and npmjs.org packages (e.g.
we publish to 2 different registries as part of the publish).

Before the change it would crash because it didn't correctly
recognize the version conflict failure coming back from github.

Now after the change, it prints the message that "all packages
have been published already" indicating that it successfully
detected all conflict errors as handled.

## Types of changes

- [ ] Chore (change that has absolutely no effect on users)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

ghiscoding commented:
Thanks for the contribution, it looks like the Lint is failing, could you please run the Prettier script or run full build which indirectly runs Prettier. Read the [Contributing](https://github.com/lerna-lite/lerna-lite/blob/main/CONTRIBUTING.md) guide for more info.

It would also be great if you could show a before and after of what this PR resolves, I'm a visual person and like to see diff if possible. Thank. I will review the code later

ghiscoding commented on packages/publish/src/lib/is-npm-js-publish-version-conflict.ts:
```diff
@@ -0,0 +1,45 @@
+    ex.message?.includes('You cannot publish over the previously published versions')
+  ) {
+    return true;
+  } else {
```
this last `else` is probably not needed since all other conditions already have a return then the only thing possible would be this `return false;`

petermetz replied:
@ghiscoding Gotcha, I'm updating it that way!

petermetz commented:
> it looks like the Lint is failing, could you please run the Prettier script or run full build which indirectly runs Prettier. 

@ghiscoding Thank you, fixed in my latest commit! I also removed the redundant `else { }` as you asked.

ghiscoding commented:
@petermetz 
Thanks the code is looking good now, I'm now looking at the tests and it seems the coverage went down and looking at codecov report, it looks like some lines from these 2 new utils were not covered. Would it be a lot of work to also cover these lines? If it's too much work, we could skip that for now but it would be nice to have good coverage of the lib. Let me know if it's possible or not. Thanks

https://app.codecov.io/gh/lerna-lite/lerna-lite/pull/738/blob/packages/publish/src/lib/is-npm-pkg-github-publish-version-conflict.ts

![image](https://github.com/lerna-lite/lerna-lite/assets/643976/1adcb6c9-d3e5-4452-a5fb-538f0f786fc0)

![image](https://github.com/lerna-lite/lerna-lite/assets/643976/131d5fe3-48b9-47c0-ade3-7827bb886b89)

petermetz commented:
> @petermetz Thanks the code is looking good now, I'm now looking at the tests and it seems the coverage went down and looking at codecov report, it looks like some lines from these 2 new utils were not covered. Would it be a lot of work to also cover these lines? If it's too much work, we could skip that for now but it would be nice to have good coverage of the lib. Let me know if it's possible or not. Thanks
> 
> https://app.codecov.io/gh/lerna-lite/lerna-lite/pull/738/blob/packages/publish/src/lib/is-npm-pkg-github-publish-version-conflict.ts
> 
> ![image](https://user-images.githubusercontent.com/643976/272687829-1adcb6c9-d3e5-4452-a5fb-538f0f786fc0.png)
> 
> ![image](https://user-images.githubusercontent.com/643976/272687984-131d5fe3-48b9-47c0-ade3-7827bb886b89.png)

@ghiscoding No worries, adding some coverage for those extra branches was relatively simple so I just got it done with my latest update!

ghiscoding commented:
> > @petermetz Thanks the code is looking good now, I'm now looking at the tests and it seems the coverage went down and looking at codecov report, it looks like some lines from these 2 new utils were not covered. Would it be a lot of work to also cover these lines? If it's too much work, we could skip that for now but it would be nice to have good coverage of the lib. Let me know if it's possible or not. Thanks
> > https://app.codecov.io/gh/lerna-lite/lerna-lite/pull/738/blob/packages/publish/src/lib/is-npm-pkg-github-publish-version-conflict.ts
> > ![image](https://user-images.githubusercontent.com/643976/272687829-1adcb6c9-d3e5-4452-a5fb-538f0f786fc0.png)
> > ![image](https://user-images.githubusercontent.com/643976/272687984-131d5fe3-48b9-47c0-ade3-7827bb886b89.png)
> 
> @ghiscoding No worries, adding some coverage for those extra branches was relatively simple so I just got it done with my latest update!

@petermetz 
hey sorry to bother again, there's still missing coverage but I think these lines could actually be merged into 1 `if` with all 3 conditions since they all return `false` anyway. That would be cheating but it would add full coverage and we probably don't care as much for the first 2 conditions anyway. I promise that's my last thing to ask 😆 

![image](https://github.com/lerna-lite/lerna-lite/assets/643976/a9e7eea7-1a05-4838-965e-7c0aeb2d2107)

petermetz commented:
> > > @petermetz Thanks the code is looking good now, I'm now looking at the tests and it seems the coverage went down and looking at codecov report, it looks like some lines from these 2 new utils were not covered. Would it be a lot of work to also cover these lines? If it's too much work, we could skip that for now but it would be nice to have good coverage of the lib. Let me know if it's possible or not. Thanks
> > > https://app.codecov.io/gh/lerna-lite/lerna-lite/pull/738/blob/packages/publish/src/lib/is-npm-pkg-github-publish-version-conflict.ts
> > > ![image](https://user-images.githubusercontent.com/643976/272687829-1adcb6c9-d3e5-4452-a5fb-538f0f786fc0.png)
> > > ![image](https://user-images.githubusercontent.com/643976/272687984-131d5fe3-48b9-47c0-ade3-7827bb886b89.png)
> > 
> > 
> > @ghiscoding No worries, adding some coverage for those extra branches was relatively simple so I just got it done with my latest update!
> 
> @petermetz hey sorry to bother again, there's still missing coverage but I think these lines could actually be merged into 1 `if` with all 3 conditions since they all return `false` anyway. That would be cheating but it would add full coverage and we probably don't care as much for the first 2 conditions anyway. I promise that's my last thing to ask 😆
> 
> Also I think I can wait to publish this code change right? I mean unless you think it's urgent, if not urgent then I would rather wait a bit more since I was waiting for `cosmiconfig` to publish a new release (the other issue you've opened), they fixed the issue but it will only be published in their next upcoming v9.0 but I don't know when he intends to publish that next major version.
> 
> ![image](https://user-images.githubusercontent.com/643976/272717925-a9e7eea7-1a05-4838-965e-7c0aeb2d2107.png)

@ghiscoding Done! No worries! Publishing it is not urgent, I've got a manual patching routine that I can use to have the new code in effect and issue our releases that way for now. Thank you for doing the work you do!

ghiscoding commented:
Thanks a lot, all good now 🚀

ghiscoding commented:
@petermetz this is now available in latest release [v2.6.0](https://github.com/lerna-lite/lerna-lite/releases/tag/v2.6.0). As for the other issue, I'm still waiting for `cosmiconfig` 9.0 to come out (no news on when that will happen) but at least there's an alternative to use 8.0 in the meantime. Thanks again for your contributions. Cheers

petermetz commented:
@ghiscoding My pleasure! Thank you very much as well!
